integrationos-snapshot/src/main.rs

```diff
@@ -0,0 +1,332 @@
+    let client = reqwest::Client::default();
+    let middleware = ClientBuilder::new(client)
+        .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+        .with(TracingMiddleware::default())
+        .build();
+    let storage = GClient::new(
+        ClientConfig {
+            http: Some(middleware),
+            ..Default::default()
+        }
+        .with_auth()
+        .await?,
+    );
+
+    let subscriber = get_subscriber("snapshot".into(), "info".into(), std::io::stdout);
+    init_subscriber(subscriber);
+
+    tracing::info!("Starting Snapshot with config:\n{config}");
+
+    let client = Client::with_uri_str(&config.db_config.control_db_url).await?;
+    let database = client.database(&config.db_config.control_db_name);
+    let snapshot: MongoStore<Event> = MongoStore::new(&database, &Store::Snapshots).await?;
+
+    let started = Started::new();
+    snapshot
+        .create_one(&Event::Started(started.clone()))
+        .await?;
+
+    let saved = save(config, &snapshot, storage, &started).await;
+
+    if let Err(e) = saved {
+        snapshot
+            .create_one(&Event::Failed(Failed::new(
+                e.to_string(),
+                started.reference(),
+            )))
+            .await?;
+
+        tracing::error!("Failed to save snapshot: {e}");
+
+        return Err(e);
+    }
+
+    tracing::info!("Snapshot saved successfully");
+
+    Ok(())
+}
+
+async fn save(
+    config: SnapshotConfig,
+    snapshot: &MongoStore<Event>,
+    storage: GClient,
+    started: &Started,
+) -> Result<()> {
+    let tmp_dir = TempDir::new()?;
+    let filter =
+        doc! { "createdAt": { "$lt": (Utc::now() - CDuration::days(30)).timestamp_millis() } };
+
+    let command = Command::new("mongodump")
+        .arg("--uri")
+        .arg(&config.db_config.control_db_url)
+        .arg("--db")
+        .arg(&config.db_config.control_db_name)
+        .arg("--collection")
+        .arg(&config.event_collection_name)
+        .arg("--query")
+        .arg(serde_json::to_string(&filter)?)
+        .arg("--out")
+        .arg(tmp_dir.path())
+        .arg("--gzip")
+        .output()?;
+
+    if !command.status.success() {
+        return Err(anyhow!("Command mongodump failed: {:?}", command));
+    }
+
+    snapshot
+        .create_one(&Event::Dumped(Dumped::new(started.reference())))
+        .await?;
+
+    let base_path = tmp_dir
+        .path()
+        .join(&config.db_config.event_db_name)
+        .join(&config.event_collection_name);
+
+    if let Err(e) = upload_file(&base_path, "bson.gz", &config, &storage).await {
+        return Err(anyhow!("Failed to upload bson file: {e}"));
+    }
+
+    if let Err(e) = upload_file(&base_path, "metadata.json.gz", &config, &storage).await {
+        return Err(anyhow!("Failed to upload json file: {e}"));
+    }
+
+    let remote_path = format!("gs://{}{}", config.gs_storage_bucket, base_path.display());
+
+    snapshot
```

Comment:
You don't need this state or the Completed, since they mean the same thing.

Reply:
Sure. Having both is indeed redundant. Will keep `Completed` only.